swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift

```diff
@@ -86,35 +63,42 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
     }
   }
 
+  // This can be called by the system, or initiated by connlib.
+  // When called by the system, we call Adapter.stop() from here.
+  // When initiated by connlib, we've already called stop() there.
   override func stopTunnel(
     with reason: NEProviderStopReason, completionHandler: @escaping () -> Void
   ) {
     logger.log("stopTunnel: Reason: \(reason)")
-    adapter?.stop(reason: reason) {
-      completionHandler()
-      #if os(macOS)
-        // HACK: This is a filthy hack to work around Apple bug 32073323
-        exit(0)
+
+    if case .authenticationCanceled = reason {
+      do {
+        // Remove the passwordReference from our configuration so that it's not used again
+        // if the app is re-launched. There's no good way to send data like this from the
+        // Network Extension to the GUI, so save it to a file for the GUI to read later.
+        try String(reason.rawValue).write(to: SharedAccess.providerStopReasonURL, atomically: true, encoding: .utf8)
```

Comment:
So with using files as IPC...

- There's no chance the GUI will read this file before the tunnel gets a chance to write it?
- There's no chance the GUI would crash and then read the file after it's stale and mis-understand what state we're in?
- Does the GUI delete the file when it reads it, to avoid processing the same event twice?

Reply:
- Not really, the sequence of events is:
  - Tunnel process writes reason to file
  - Tunnel process exits
  - GUI receives `NEVPNStatusDidChange`
  - GUI reads the file and acts accordingly, deleting it
- Hm, it could, but this file is solely for (at this point) showing a UI notification. It's a boolean flag that tells it "You were unexpectedly signed out, sign in again". If the GUI crashes before it's able to "consume" the notification, it'll just show it again on next launch, no biggie (other than an annoyance to the user).
- Yes